File: typescript/src/utils/base58.ts
/**
 * Base58 encoding utilities for Solana addresses
 * 
 * Solana uses Base58 encoding (Bitcoin-style) for public keys and signatures.
 */

const BASE58_ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

/**
 * Encode bytes to Base58 string (Solana address format)
 * 
 * @param bytes - Uint8Array to encode
 * @returns Base58 encoded string
 * 
 * @example
 * ```typescript
 * const pubkey = new Uint8Array([1, 2, 3, ...]);
 * const address = base58Encode(pubkey);
 * // "EPjFWdd5..."
 * ```
 */
export function base58Encode(bytes: Uint8Array): string {
  const digits = [0];
  
  for (let i = 0; i < bytes.length; i++) {
    let carry = bytes[i];
    for (let j = 0; j < digits.length; j++) {
      carry += digits[j] << 8;
      digits[j] = carry % 58;
      carry = (carry / 58) | 0;
    }
    while (carry > 0) {
      digits.push(carry % 58);
      carry = (carry / 58) | 0;
    }
  }

  // Leading zeros
  let leadingZeros = '';
  for (let i = 0; i < bytes.length && bytes[i] === 0; i++) {
    leadingZeros += '1';
  }

  return leadingZeros + digits.reverse().map(d => BASE58_ALPHABET[d]).join('');
}

/**
 * Decode Base58 string to bytes
 * 
 * @param str - Base58 encoded string
 * @returns Uint8Array of decoded bytes
 * @throws Error if string contains invalid characters
 * 
 * @example
 * ```typescript
 * const bytes = base58Decode("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v");
 * // Uint8Array(32) [...]
 * ```
 */
export function base58Decode(str: string): Uint8Array {
  const bytes = [0];
  
  for (let i = 0; i < str.length; i++) {
    const char = str[i];
    const value = BASE58_ALPHABET.indexOf(char);
    
    if (value === -1) {
      throw new Error(`Invalid Base58 character: ${char}`);
    }
    
    let carry = value;
    for (let j = 0; j < bytes.length; j++) {
      carry += bytes[j] * 58;
      bytes[j] = carry & 0xff;
      carry >>= 8;
    }
    while (carry > 0) {
      bytes.push(carry & 0xff);
      carry >>= 8;
    }
  }

  // Leading '1's are leading zeros
  for (let i = 0; i < str.length && str[i] === '1'; i++) {
    bytes.push(0);
  }

  return new Uint8Array(bytes.reverse());
}

/**
 * Check if a string is a valid Solana public key (Base58, 32-44 chars)
 * 
 * @param address - String to validate
 * @returns true if valid Solana pubkey format
 * 
 * @example
 * ```typescript
 * isValidPubkey("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v") // true
 * isValidPubkey("invalid") // false
 * ```
 */
export function isValidPubkey(address: string): boolean {
  // Solana pubkeys are 32-44 characters in Base58
  if (address.length < 32 || address.length > 44) {
    return false;
  }

  // Check all characters are valid Base58
  for (const char of address) {
    if (!BASE58_ALPHABET.includes(char)) {
      return false;
    }
  }

  // Try to decode and verify length
  try {
    const bytes = base58Decode(address);
    return bytes.length === 32;
  } catch {
    return false;
  }
}

File: rust/src/utils/base58.rs
//! Base58 encoding/decoding utilities for Solana addresses.

/// Encode bytes to base58 string.
pub fn base58_encode(data: &[u8]) -> String {
    bs58::encode(data).into_string()
}

/// Decode base58 string to bytes.
pub fn base58_decode(s: &str) -> Result<Vec<u8>, bs58::decode::Error> {
    bs58::decode(s).into_vec()
}

/// Check if a string is a valid Solana public key.
///
/// # Arguments
///
/// * `address` - Base58-encoded address to validate
///
/// # Returns
///
/// True if valid, false otherwise
pub fn is_valid_pubkey(address: &str) -> bool {
    // Solana pubkeys are 32 bytes, which encode to 32-44 base58 chars
    if address.len() < 32 || address.len() > 44 {
        return false;
    }

    match bs58::decode(address).into_vec() {
        Ok(decoded) => decoded.len() == 32,
        Err(_) => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_valid_pubkey() {
        // Valid Solana pubkey (SOL mint)
        assert!(is_valid_pubkey(
            "So11111111111111111111111111111111111111112"
        ));
        // Valid USDC mint
        assert!(is_valid_pubkey(
            "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"
        ));
    }

    #[test]
    fn test_invalid_pubkey() {
        assert!(!is_valid_pubkey(""));
        assert!(!is_valid_pubkey("tooshort"));
        assert!(!is_valid_pubkey("invalid!@#$"));
    }

    #[test]
    fn test_base58_roundtrip() {
        let original = vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
        let encoded = base58_encode(&original);
        let decoded = base58_decode(&encoded).unwrap();
        assert_eq!(original, decoded);
    }
}
